pkg/cloud/api/mutable_resource.go

```diff
@@ -272,24 +291,51 @@ func (u *mutableResource[GA, Alpha, Beta]) AccessBeta(f func(x *Beta)) error {
 	return u.postAccess(meta.VersionBeta, 0)
 }
 
+// completeType means that the object does not have conversion errors and its
+// type is not missing.
+func completeType(err error, obj interface{}) bool {
+	return err == nil && !IsTypeMissing(obj)
+}
+
+// convertibleTo means that the object does not have conversion errors or its
+// type is missing.
+func convertibleTo(err error, obj interface{}) bool {
+	return err == nil || IsTypeMissing(obj)
+}
+
+// convertibleTo means that the object has conversion errors or its type is
+// missing.
+func notConvertibleTo(err error, obj interface{}) bool {
+	return err != nil || IsTypeMissing(obj)
+}
+
 func (u *mutableResource[GA, Alpha, Beta]) ImpliedVersion() (meta.Version, error) {
 	_, gaErr := u.ToGA()
 	_, alphaErr := u.ToAlpha()
 	_, betaErr := u.ToBeta()
 
-	switch {
-	case gaErr == nil && alphaErr == nil && betaErr == nil:
+	// If type is GA and converts to Alpha and Beta without error return GA.
```

Comment:
It seems like the code you have here is redundant (I might be missing something)

`ToGA`, `ToAlpha`, `ToBeta` already return an error when the type that is being converted to is `PlaceholderType`. This means that we don't have to have an additional level of checking to implement the logic table above.

To make the code match the logic table almost exactly, I recommend something like this:

```
	type v struct{ ga, alpha, beta bool }
	result := v{ga: gaErr != nil, alpha: alphaErr != nil, beta: betaErr != nil}
	switch result {
	case v{true, true, true}:
    return meta.VersionGA, nil
	case v{true, true, false}:
    return meta.VersionAlpha, nil
		//...
	}
```

Reply:
I added those functions for better understanding why we imply such version like `GA needs to be complete type and should convert to Alpha and Beta`. where complete type means that GA type cannot be `PlaceholderType` and convert to means that there is no error from conversion or type is `PlaceholderType`. It is more readable for me :) thank `err != nil` 

In your proposed solution we are missing the information why conversion failed.
For GA we should return GA when Alpha and Beta convert without errors or are `PlaceholderType` in other case where conversion fails due to some other reason we should return error.

Reply:
I find that the tabular format is easier to understand. 

In your code below, I have to understand both what convertibleTo does, the first argument is a generic error type (that could be anything) and the second is a type that is passed to a typecheck. It's non-trivial to check that the behavior given above is matching the code (you can do it by inspection) rather than having to reason about the `if` and `switch` statements below.

Reply:
I don't mind adding tabular format (when we agree on the result) I want also to add functions that are descriptive.

Reply:
The discussion was moved offline. Resolving this thread